Update course

LuisRobles18 opened the pull request.

Commit:
Added script SQL for course table

Commit:
Added model layer for Courses

Commit:
Implemented all layers for Course Management Module (Model, Service,
Controller, Repository)

Commit:
All validations implemented to create the course.

Commit:
Added the intermediate entity for Categories and Courses

Commit:
Finished the functionallity for create courses module.

Commit:
Added Swagger documentation for create course module

Commit:
Added jUnit tests for create course module

Commit:
Implemented all functionallities for update courses module.

Commit:
Added jUnit testings for update course module.

Commit:
Fixed jUnit testings for update course module

Commit:
Fixed jUnit testings for update course module

Commit:
Branch ready to be reviewed

angel5644 commented on User Management/scripts/catcourses.sql:
```diff
@@ -0,0 +1,39 @@
+  STORAGE(INITIAL 65536 NEXT 1048576 MINEXTENTS 1 MAXEXTENTS 2147483645
+  PCTINCREASE 0 FREELISTS 1 FREELIST GROUPS 1 BUFFER_POOL DEFAULT FLASH_CACHE DEFAULT CELL_FLASH_CACHE DEFAULT)
+  TABLESPACE "SYSTEM"  ENABLE;
+  ALTER TABLE "TRAININGTRACKING"."CATCOURSES" MODIFY ("IDCOURSE" NOT NULL ENABLE);
```
I think we are missing the FK for IDCOURSE and IDCATEGORY

angel5644 commented on User Management/scripts/courses.sql:
```diff
@@ -0,0 +1,59 @@
+   (	"ID" NUMBER(*,0), 
+	"NAME" VARCHAR2(50 BYTE), 
+	"DESCRIPTION" VARCHAR2(500 BYTE), 
+	"CATEGORY" NUMBER(*,0), 
```
Due to a course can have more than one category. There is no need to have a field for "Category", the categories related to the course would be included in a different table.

angel5644 commented on User Management/src/main/java/com/usermanagement/model/Courses.java:
```diff
@@ -0,0 +1,32 @@
+@Entity
+@Table(name = "COURSES", catalog = "XE")
+@Data
+public class Courses {
```
When modeling an individual model, we should use singular names. This model would store an individual course, therefore the proper name for this one may be: "Course". "Courses" for the table its fine.

angel5644 commented on User Management/src/main/java/com/usermanagement/repository/CatCourseRepository.java:
```diff
@@ -0,0 +1,15 @@
+	@Transactional
+	@Modifying
+	@Query("DELETE FROM CatCourses c WHERE idcourse = ?1")
+	void deleteCategoriesByCourse(Integer id);
```
As a good practice, we should try to return always a value so unit testing can be implemented in an easier way. A boolean data type may be used, returning a true if the course was deleted from the repository.

angel5644 commented on User Management/src/main/java/com/usermanagement/model/CatCourses.java:
```diff
@@ -0,0 +1,30 @@
+	@Column(name = "id",length=32, unique = true, nullable = false, insertable = false, updatable = false)
+	@GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "SEQUENCE_NAME_CATCOURSE")
+	@SequenceGenerator(name = "SEQUENCE_NAME_CATCOURSE", sequenceName = "SEQUENCE_NAME_CATCOURSE", allocationSize = 1, initialValue = 1)
+	private Integer id;
```
I would stick to the primitive type "int" in this case, instead of the class which is a reference type. Take into account that if any calculations are made, having reference types instead of a primitive may cause performance issues.

angel5644 commented on User Management/src/main/java/com/usermanagement/service/CourseManager.java:
```diff
@@ -0,0 +1,14 @@
+
+public interface CourseManager {
+
+	public Boolean createCourse(CourseRequest theCourse);
```
In my opinion, here we should be returning the CreateCourseResponse object instead of a boolean. The same with update, probably we would return UpdateCourseResponse.

angel5644 commented on User Management/src/main/java/com/usermanagement/service/CourseManager.java:
```diff
@@ -0,0 +1,14 @@
+
+	public Boolean createCourse(CourseRequest theCourse);
+	public Boolean updateCourse(Integer id, CourseRequest theCourse);
+	public Boolean validateFields(CourseRequest theCourse);
```
Let's stick to the primitive types (boolean instead of Boolean) instead of reference types.

angel5644 commented on User Management/src/main/java/com/usermanagement/service/CourseManagerImpl.java:
```diff
@@ -0,0 +1,318 @@
+	@Transactional
+	public Boolean createCourse(CourseRequest theCourse) throws DuplicateKeyException {
+		result = "";
+		Boolean isOk = true;
```
Same as in other observations, for operations and calculations use primitive types if possible.

angel5644 commented on User Management/src/main/java/com/usermanagement/service/CourseManagerImpl.java:
```diff
@@ -0,0 +1,318 @@
+		return isOk;
+	}
+
+	public Boolean validateFields(CourseRequest theCourse) throws InvalidParameterException {
```
I like the way of handling business validations. (y)

angel5644 commented on User Management/src/test/java/com/usermanagement/Course/CourseManagement/CreateCourseTests.java:
```diff
@@ -0,0 +1,197 @@
+	@Autowired
+	CourseResource courseResource;
+	
+	/*Test cases
```
Nice comments.

angel5644 commented on User Management/src/test/java/com/usermanagement/Course/CourseManagement/CreateCourseTests.java:
```diff
@@ -0,0 +1,197 @@
+	}
+	
+	@Test
+	public void insertCourseSuccessfullyWithEmptyDescription() {
```
Good descriptive test name.

angel5644 commented on User Management/src/test/java/com/usermanagement/Course/CourseManagement/CreateCourseTests.java:
```diff
@@ -0,0 +1,197 @@
+	}
+	
+	@Test
+	public void insertCourseWithEmptyCategories() {
```
Nice way of organizing. I prefer using the AAA pattern, separating in arrange, act and assert the test. This way the test is a little bit more structured and easy to understand.

angel5644 reviewed (changes requested):
Please check my comments.

Commit:
Changes made according to reviewer's feedback from Pull request #7.

IvanVegaS commented on User Management/src/main/java/com/usermanagement/model/Category.java:
```diff
@@ -13,19 +13,19 @@
 	private String description;
 	
-	public Categories(){}
+	public Category(){}
```
If lombok is already being used, I recommend using the annotations @NoArgsConstructor and @AllArgsConstructor instead of manually declaring it.

IvanVegaS commented on User Management/src/main/java/com/usermanagement/model/Course.java:
```diff
@@ -0,0 +1,32 @@
+	@Column(name = "content",nullable = false, length = 4000)
+	private String content;
+	
+	public Course(){}
```
If lombok is already being used, I recommend using the annotations @NoArgsConstructor and @AllArgsConstructor instead of manually declaring it.

angel5644 replied:
@IvanVegaS, could you check and aprove changes, please?

Commit:
Added the @NoArgsConstructor annotation for each entity that has a
manually declared constructor. (Also removed their manually declared
constructors)

LuisRobles18 commented:
I´ve added and modified every comments from Víctor and Ivan.
